syntax = "proto3";

package rtype.auth;

// Types de messages d'authentification
enum AuthMessageType {
    AUTH_UNKNOWN = 0;
    AUTH_LOGIN_REQUEST = 1;
    AUTH_LOGIN_RESPONSE = 2;
    AUTH_REGISTER_REQUEST = 3;
    AUTH_REGISTER_RESPONSE = 4;
    AUTH_LOGOUT_REQUEST = 5;
    AUTH_LOGOUT_RESPONSE = 6;
    AUTH_TOKEN_REFRESH = 7;
}

// Codes de réponse
enum AuthResponseCode {
    AUTH_SUCCESS = 0;
    AUTH_INVALID_CREDENTIALS = 1;
    AUTH_USER_NOT_FOUND = 2;
    AUTH_USER_ALREADY_EXISTS = 3;
    AUTH_INVALID_TOKEN = 4;
    AUTH_TOKEN_EXPIRED = 5;
    AUTH_SERVER_ERROR = 6;
    AUTH_INVALID_FORMAT = 7;
}

// Requête de connexion
message LoginRequest {
    string username = 1;
    string password_hash = 2;  // SHA-256 hash du mot de passe
    string client_version = 3;
}

// Réponse de connexion
message LoginResponse {
    AuthResponseCode code = 1;
    string message = 2;
    string session_token = 3;
    string player_id = 4;
    uint64 token_expires_at = 5;  // Unix timestamp
}

// Requête d'inscription
message RegisterRequest {
    string username = 1;
    string email = 2;
    string password_hash = 3;
}

// Réponse d'inscription
message RegisterResponse {
    AuthResponseCode code = 1;
    string message = 2;
    string player_id = 3;
}

// Message d'authentification générique (wrapper)
message AuthMessage {
    AuthMessageType type = 1;
    uint32 sequence_id = 2;
    uint64 timestamp_ms = 3;

    oneof payload {
        LoginRequest login_request = 10;
        LoginResponse login_response = 11;
        RegisterRequest register_request = 12;
        RegisterResponse register_response = 13;
    }
}
